File: graph_components/dfs/src/DfsDemo.cs
// C#
// グラフの連結成分: DFS

using System;
using System.Collections.Generic;
using System.Linq;

public class GraphData
{
    private Dictionary<string, List<(string Neighbor, int Weight)>> _data;

    public GraphData()
    {
        _data = new Dictionary<string, List<(string, int)>>();
    }

    public Dictionary<string, List<(string, int)>> Get()
    {
        return _data;
    }

    public List<string> GetVertices()
    {
        return _data.Keys.ToList();
    }

    public bool AddVertex(string vertex)
    {
        if (!_data.ContainsKey(vertex))
        {
            _data[vertex] = new List<(string, int)>();
        }
        return true;
    }

    public bool AddEdge(string vertex1, string vertex2, int weight)
    {
        AddVertex(vertex1);
        AddVertex(vertex2);

        // Ensure no duplicate edges for vertex1 -> vertex2
        var existsV1V2 = _data[vertex1].FindIndex(n => n.Neighbor == vertex2);
        if (existsV1V2 != -1)
        {
            _data[vertex1][existsV1V2] = (vertex2, weight);
        }
        else
        {
            _data[vertex1].Add((vertex2, weight));
        }

        // Ensure no duplicate edges for vertex2 -> vertex1
        var existsV2V1 = _data[vertex2].FindIndex(n => n.Neighbor == vertex1);
        if (existsV2V1 != -1)
        {
            _data[vertex2][existsV2V1] = (vertex1, weight);
        }
        else
        {
            _data[vertex2].Add((vertex1, weight));
        }

        return true;
    }

    public bool Clear()
    {
        _data.Clear();
        return true;
    }

    private void Dfs(string vertex, HashSet<string> visited, List<string> currentComponent)
    {
        visited.Add(vertex);
        currentComponent.Add(vertex);

        if (_data.TryGetValue(vertex, out var neighbors))
        {
            foreach (var (neighborVertex, _) in neighbors)
            {
                if (!visited.Contains(neighborVertex))
                {
                    Dfs(neighborVertex, visited, currentComponent);
                }
            }
        }
    }

    public List<List<string>> GetConnectedComponents()
    {
        var visited = new HashSet<string>();
        var connectedComponents = new List<List<string>>();

        foreach (var vertex in GetVertices())
        {
            if (!visited.Contains(vertex))
            {
                var currentComponent = new List<string>();
                Dfs(vertex, visited, currentComponent);
                connectedComponents.Add(currentComponent);
            }
        }

        return connectedComponents;
    }

    public static void Main(string[] args)
    {
        Console.WriteLine("Dfs TEST -----> start");

        Console.WriteLine("\nnew");
        var graphData = new GraphData();
        Console.WriteLine($"  現在のデータ: {string.Join(", ", graphData.Get())}");

        Console.WriteLine("\nadd_edge");
        graphData.Clear();
        var inputList = new[]
        {
            ("A", "B", 4), ("B", "C", 3), ("B", "D", 2), 
            ("D", "A", 1), ("A", "C", 2), ("B", "D", 2)
        };
        
        foreach (var input in inputList)
        {
            Console.WriteLine($"  入力値: {input}");
            var output = graphData.AddEdge(input.Item1, input.Item2, input.Item3);
            Console.WriteLine($"  出力値: {output}");
        }
        Console.WriteLine($"  現在のデータ: {string.Join(", ", graphData.Get())}");
        
        Console.WriteLine("\nget_connected_components");
        var connectedComponents = graphData.GetConnectedComponents();
        Console.WriteLine($"  連結成分: {string.Join(", ", connectedComponents.Select(c => string.Join(",", c)))}");

        // Similar code blocks for other test cases can be added here
        Console.WriteLine("Dfs TEST <----- end");
    }
}

File: graph_components/dfs/src/DfsDemo.rs
// Rust
// グラフの連結成分: DFS

use std::collections::{HashMap, HashSet};

struct GraphData {
    // 隣接ノードとその辺の重みを格納
    // キーは頂点、値はその頂点に隣接する頂点と重みのベクター
    data: HashMap<String, Vec<(String, i32)>>,
}

impl GraphData {
    fn new() -> Self {
        GraphData {
            data: HashMap::new(),
        }
    }

    fn get(&self) -> &HashMap<String, Vec<(String, i32)>> {
        // グラフの内部データを取得
        &self.data
    }

    fn get_vertices(&self) -> Vec<String> {
        // グラフの全頂点をベクトルとして返す
        self.data.keys().cloned().collect()
    }

    fn add_vertex(&mut self, vertex: &str) -> bool {
        // 新しい頂点をグラフに追加
        self.data.entry(vertex.to_string()).or_insert_with(Vec::new);
        true
    }
    
    fn add_edge(&mut self, vertex1: &str, vertex2: &str, weight: i32) -> bool {
        // 両頂点間に辺を追加。重みを指定
        // 頂点がグラフに存在しない場合は追加
        
        // 頂点を確実に追加
        self.add_vertex(vertex1);
        self.add_vertex(vertex2);
        
        // vertex1 -> vertex2 の辺を追加（重み付き）
        let mut edge_exists_v1v2 = false;
        if let Some(neighbors) = self.data.get_mut(vertex1) {
            for i in 0..neighbors.len() {
                if neighbors[i].0 == vertex2 {
                    neighbors[i].1 = weight; // 既に存在する場合は重みを更新
                    edge_exists_v1v2 = true;
                    break;
                }
            }
            if !edge_exists_v1v2 {
                neighbors.push((vertex2.to_string(), weight));
            }
        }
        
        // vertex2 -> vertex1 の辺を追加（重み付き）
        let mut edge_exists_v2v1 = false;
        if let Some(neighbors) = self.data.get_mut(vertex2) {
            for i in 0..neighbors.len() {
                if neighbors[i].0 == vertex1 {
                    neighbors[i].1 = weight; // 既に存在する場合は重みを更新
                    edge_exists_v2v1 = true;
                    break;
                }
            }
            if !edge_exists_v2v1 {
                neighbors.push((vertex1.to_string(), weight));
            }
        }
        
        true
    }
        
    fn clear(&mut self) -> bool {
        // グラフを空にする
        self.data.clear();
        true
    }
    
    fn _dfs(&self, vertex: &str, visited: &mut HashSet<String>, current_component: &mut Vec<String>) {
        // 現在の頂点を訪問済みにマークし、現在の成分に追加
        visited.insert(vertex.to_string());
        current_component.push(vertex.to_string());
        
        // 隣接する頂点を探索
        if let Some(neighbors) = self.data.get(vertex) {
            for (neighbor_vertex, _) in neighbors {
                // まだ訪問していない隣接頂点に対して再帰的にDFSを呼び出す
                if !visited.contains(neighbor_vertex) {
                    self._dfs(neighbor_vertex, visited, current_component);
                }
            }
        }
    }
    
    fn get_connected_components(&self) -> Vec<Vec<String>> {
        let mut visited = HashSet::new(); // 訪問済み頂点を記録するセット
        let mut connected_components = Vec::new(); // 連結成分を格納するベクトル
        
        // グラフの全頂点を順にチェック
        for vertex in self.get_vertices() {
            // まだ訪問していない頂点からDFSを開始
            if !visited.contains(&vertex) {
                let mut current_component = Vec::new(); // 現在の連結成分を格納するベクトル
                // DFSヘルパー関数を呼び出し、現在の連結成分を探索
                self._dfs(&vertex, &mut visited, &mut current_component);
                // 探索で見つかった連結成分を結果ベクトルに追加
                connected_components.push(current_component);
            }
        }
        
        connected_components
    }
}

fn main() {
    println!("Dfs TEST -----> start");
    
    println!("\nnew");
    let mut graph_data = GraphData::new();
    println!("  現在のデータ: {:?}", graph_data.get());
    
    println!("\nadd_edge");
    graph_data.clear();
    let input_list = vec![
        ("A", "B", 4), ("B", "C", 3), ("B", "D", 2), 
        ("D", "A", 1), ("A", "C", 2), ("B", "D", 2)
    ];
    for input in input_list {
        println!("  入力値: {:?}", input);
        let output = graph_data.add_edge(input.0, input.1, input.2);
        println!("  出力値: {}", output);
    }
    println!("  現在のデータ: {:?}", graph_data.get());
    println!("\nget_connected_components");
    let output = graph_data.get_connected_components();
    println!("  連結成分: {:?}", output);
    
    println!("\nadd_edge");
    graph_data.clear();
    let input_list = vec![
        ("A", "B", 4), ("C", "D", 4), ("E", "F", 1), ("F", "G", 1)
    ];
    for input in input_list {
        println!("  入力値: {:?}", input);
        let output = graph_data.add_edge(input.0, input.1, input.2);
        println!("  出力値: {}", output);
    }
    println!("  現在のデータ: {:?}", graph_data.get());
    println!("\nget_connected_components");
    let output = graph_data.get_connected_components();
    println!("  連結成分: {:?}", output);
    
    println!("\nadd_edge");
    graph_data.clear();        
    let input_list = vec![("A", "B", 4), ("B", "C", 3), ("D", "E", 5)];
    for input in input_list {
        println!("  入力値: {:?}", input);
        let output = graph_data.add_edge(input.0, input.1, input.2);
        println!("  出力値: {}", output);
    }
    println!("  現在のデータ: {:?}", graph_data.get());
    println!("\nget_connected_components");
    let output = graph_data.get_connected_components();
    println!("  連結成分: {:?}", output);
    
    println!("\nadd_edge");
    graph_data.clear();
    let input_list: Vec<(&str, &str, i32)> = vec![];
    for input in input_list {
        println!("  入力値: {:?}", input);
        let output = graph_data.add_edge(input.0, input.1, input.2);
        println!("  出力値: {}", output);
    }
    println!("  現在のデータ: {:?}", graph_data.get());
    println!("\nget_connected_components");
    let output = graph_data.get_connected_components();
    println!("  連結成分: {:?}", output);
    
    println!("Dfs TEST <----- end");
}